Feature/create user

gogohaena opened the pull request:
## #️⃣연관된 이슈

Close #10 
Close #17 
Close #19 
Close #20 

## 📝작업 내용

- 회원가입, 사용자 등록 기능 구현
- 로그인 (jwt 반환) 기능 구현

### 스크린샷 (선택)
회원가입
```
curl --location 'http://localhost:19093/auth/signup' \
--header 'Content-Type: application/json' \
--data-raw '{
    "username" : "testuser1",
    "nickname" : "TestNickname1",
    "email" : "testuser1@example.com",
    "password" : "Password@123",
    "role" : "COMPANY_USER"  
}'
```
로그인
```
curl --location 'http://localhost:19093/auth/signin' \
--header 'Content-Type: application/json' \
--data-raw '{
    "username" : "testuser1",
    "password" : "Password@123"
}'
```

## 💬리뷰 요구사항(선택)

>

gogohaena committed:
Feat: Libs 추가

- BaseEntity
- GlobalException
- Error/SuccessResponse

gogohaena committed:
Feat: User Entity 추가

- User, UserRole
- UserRepository, UserRepositoryImpl

gogohaena committed:
Merge branch 'develop' into feature/create-user

# Conflicts:
#	.gitignore

gogohaena committed:
Feat: PostgreSQL 설정 추가

gogohaena committed:
Fix: 의존성 수정

gogohaena committed:
Feat: User DTO 작성 및 User 엔티티 수정

- UserRequestDto 작성 : validation 체크
- UserResponseDto 작성
- User 엔티티 수정 : validation을 DTO에서 검증
- UserTest 작성

gogohaena committed:
Refactor: import 정리

gogohaena committed:
Feat: 사용자 등록 기능 구현

gogohaena committed:
Feat: 회원가입 기능 구현, User id (Long) 추가

gogohaena committed:
Feat: 로그인 기능 (액세스 토큰 반환) 구현

gogohaena committed:
Refactor: import 정리

handyejee commented on com.sparta.logistics.auth/src/main/java/com/sparta/logistics/auth/domain/model/User.java:
```diff
@@ -0,0 +1,77 @@
+@Entity
+@Table(name = "p_user")
+@Getter
+@Builder(access = AccessLevel.PRIVATE)
```
Builder 어노테이션에서도 access level 옵션을 사용할 수 있는지 처음 알았어요 ! 외부 접근 제어할때 고려해 볼 수 있을것 같아요 :)

handyejee commented on com.sparta.logistics.auth/src/main/java/com/sparta/logistics/auth/domain/model/UserRole.java:
```diff
@@ -0,0 +1,21 @@
+    DELIVERY_USER("배송 담당자"),
+    COMPANY_USER("업체 담당자");
+
+    private final String description;
```
여기서 description은 어떤 역할을 하는지 궁금합니다! exception 에 있는 description이랑 같은 항목일까요?

gogohaena replied:
딱히 사용처는 없을 것 같은데 정보성으로 발제문에 나온 키워드대로 명시해주고싶어서 적어봤읍니다,,

handyejee commented on com.sparta.logistics.auth/src/main/java/com/sparta/logistics/auth/libs/model/BaseEntity.java:
```diff
@@ -0,0 +1,48 @@
+    private LocalDateTime updatedAt;
+
+    @LastModifiedBy
+    @Column(length = 100)
```
저도 수정한 부분인데, 테이블과의 매핑이 안되어 있어 Column 어노테이션에서 `name = created_at` 과 같이 수정이 필요해 보입니다 !

gogohaena replied:
확인해보겠습니다

yoaruku replied:
카멜케이스 기법을 이용해 변수명을 선언하면 JPA가 DB설정할때 자동으로 대문자앞에 `_` 룰 붙여주는걸로 알고있었는데 저도 수정해야할까요??

geonuu09 replied:
저도 그렇게 알고 있습니다! 그래서 컬럼 지정안했는데.. 내일 스크럼때 정해볼까요?

handyejee commented on com.sparta.logistics.auth/src/main/java/com/sparta/logistics/auth/presentation/controller/AuthController.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    // username 존재 여부 확인
+    @GetMapping("/verify-username")
```
이 부분은 제가 잘몰라서 그러는데, 이메일, 사용자 존재 여부를 확인하기 위해 내부 서비스에서 API를 호출해야 하는 이유가 궁금합니다 ! 필요한 경우라면 API 명세서에는 없어서 추가가 필요해 보입니다. 마지막으로 저의 작은 견해로는 URI 에는 행위를 의미하는 동사가 들어가는 것을 지양하는 것이 좋다고 알고 있어 꼭 필요한 경우가 아니라면 명사로 변경해보는것도 제안드립니다 :)

gogohaena replied:
UX/UI 관점에서 일반적으로 회원가입 시에 사용자 ID나 email 중복검사를 프론트엔드에서 ajax로 호출하기 때문에 API를 추가했습니다. API 명세서에도 추가해두도록 하겠습니다 :) URI는 명사로 바꾸는 것 좋은 것 같습니다 ㅎㅎ 감사합니다

gogohaena committed:
Refactor: 회원가입 관련 validation API 수정

- username, email, slackId 검증 엔드포인트 수정 : GET /auth/validation
- BaseEntity에 컬럼네임 명시

yoaruku reviewed (approved):
고생하셨습니다

geonuu09 reviewed (approved):
적용해봐도 좋을 것 같습니다!
